### ISSUE TITLE
Bug: Weird output from CodeQwen converted from safetensors and unrecognized BPE pre-tokenizer for CodeQwen

### ISSUE THREAD

**Issue by jarca0123** [ggerganov/llama.cpp#7939]
### What happened?

When trying to convert CodeQwen safetensors into GGUF, I get this error:
```
python3.10 convert-hf-to-gguf.py ./models--Qwen--CodeQwen1.5-7B-Chat/snapshots/7b0cc3380fe815e6f08fe2f80c03e05a8b1883d8/ --outfile test.gguf 

INFO:hf-to-gguf:Loading model: 7b0cc3380fe815e6f08fe2f80c03e05a8b1883d8
INFO:gguf.gguf_writer:gguf: This GGUF file is for Little Endian only
INFO:hf-to-gguf:Set model parameters
INFO:hf-to-gguf:gguf: context length = 65536
INFO:hf-to-gguf:gguf: embedding length = 4096
INFO:hf-to-gguf:gguf: feed forward length = 13440
INFO:hf-to-gguf:gguf: head count = 32
INFO:hf-to-gguf:gguf: key-value head count = 4
INFO:hf-to-gguf:gguf: rope theta = 1000000
INFO:hf-to-gguf:gguf: rms norm epsilon = 1e-05
INFO:hf-to-gguf:gguf: file type = 1
INFO:hf-to-gguf:Set model tokenizer
WARNING:hf-to-gguf:

WARNING:hf-to-gguf:**************************************************************************************
WARNING:hf-to-gguf:** WARNING: The BPE pre-tokenizer was not recognized!
WARNING:hf-to-gguf:**          There are 2 possible reasons for this:
WARNING:hf-to-gguf:**          - the model has not been added to convert-hf-to-gguf-update.py yet
WARNING:hf-to-gguf:**          - the pre-tokenization config has changed upstream
WARNING:hf-to-gguf:**          Check your model files and convert-hf-to-gguf-update.py and update them accordingly.
WARNING:hf-to-gguf:** ref:     https://github.com/ggerganov/llama.cpp/pull/6920
WARNING:hf-to-gguf:**
WARNING:hf-to-gguf:** chkhsh:  cd88bc280b3debbdddec3304015afd7e215c61d674846a2dac7271275384810c
WARNING:hf-to-gguf:**************************************************************************************
WARNING:hf-to-gguf:

Traceback (most recent call last):
  File "./llama.cpp/convert-hf-to-gguf.py", line 1622, in set_vocab
    self._set_vocab_sentencepiece()
  File "./llama.cpp/convert-hf-to-gguf.py", line 582, in _set_vocab_sentencepiece
    raise FileNotFoundError(f"File not found: {tokenizer_path}")
FileNotFoundError: File not found: ./models--Qwen--CodeQwen1.5-7B-Chat/snapshots/7b0cc3380fe815e6f08fe2f80c03e05a8b1883d8/tokenizer.model

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./llama.cpp/convert-hf-to-gguf.py", line 2881, in <module>
    main()
  File "./llama.cpp/convert-hf-to-gguf.py", line 2866, in main
    model_instance.set_vocab()
  File "./llama.cpp/convert-hf-to-gguf.py", line 1624, in set_vocab
    self._set_vocab_gpt2()
  File "./llama.cpp/convert-hf-to-gguf.py", line 509, in _set_vocab_gpt2
    tokens, toktypes, tokpre = self.get_vocab_base()
  File "./llama.cpp/convert-hf-to-gguf.py", line 382, in get_vocab_base
    tokpre = self.get_vocab_base_pre(tokenizer)
  File "./llama.cpp/convert-hf-to-gguf.py", line 500, in get_vocab_base_pre
    raise NotImplementedError("BPE pre-tokenizer was not recognized - update get_vocab_base_pre()")
NotImplementedError: BPE pre-tokenizer was not recognized - update get_vocab_base_pre()
```
When replacing the chkhsh from this (codeqwen) to qwen2 (or by adding codeqwen into convert-hf-to-gguf-update.py and adding new lines to llama.cpp and llama.h corresponding to qwen2), it works, and quantization also works, however, the output is with some unknown bytes. I've named the quantized model "testbase".
```
Log start
main: build = 3148 (e65bbf60)
main: built with cc (GCC) 14.1.1 20240522 for x86_64-pc-linux-gnu
main: seed  = 1718375838
llama_model_loader: loaded meta data with 24 key-value pairs and 387 tensors from ./testbase (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = qwen2
llama_model_loader: - kv   1:                               general.name str              = 7b0cc3380fe815e6f08fe2f80c03e05a8b1883d8
llama_model_loader: - kv   2:                          qwen2.block_count u32              = 32
llama_model_loader: - kv   3:                       qwen2.context_length u32              = 65536
llama_model_loader: - kv   4:                     qwen2.embedding_length u32              = 4096
llama_model_loader: - kv   5:                  qwen2.feed_forward_length u32              = 13440
llama_model_loader: - kv   6:                 qwen2.attention.head_count u32              = 32
llama_model_loader: - kv   7:              qwen2.attention.head_count_kv u32              = 4
llama_model_loader: - kv   8:                       qwen2.rope.freq_base f32              = 1000000,000000
llama_model_loader: - kv   9:     qwen2.attention.layer_norm_rms_epsilon f32              = 0,000010
llama_model_loader: - kv  10:                          general.file_type u32              = 2
llama_model_loader: - kv  11:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  12:                         tokenizer.ggml.pre str              = codeqwen
llama_model_loader: - kv  13:                      tokenizer.ggml.tokens arr[str,92416]   = ["<unk>", "<s>", "<|endoftext|>", "<|...
llama_model_loader: - kv  14:                  tokenizer.ggml.token_type arr[i32,92416]   = [3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, ...
llama_model_loader: - kv  15:                      tokenizer.ggml.merges arr[str,112005]  = ["▁ t", "▁ a", "i n", "h e", "e r...
llama_model_loader: - kv  16:                tokenizer.ggml.bos_token_id u32              = 2
llama_model_loader: - kv  17:                tokenizer.ggml.eos_token_id u32              = 4
llama_model_loader: - kv  18:            tokenizer.ggml.unknown_token_id u32              = 0
llama_model_loader: - kv  19:            tokenizer.ggml.padding_token_id u32              = 92298
llama_model_loader: - kv  20:               tokenizer.ggml.add_bos_token bool             = false
llama_model_loader: - kv  21:               tokenizer.ggml.add_eos_token bool             = false
llama_model_loader: - kv  22:                    tokenizer.chat_template str              = {% for message in messages %}{% if lo...
llama_model_loader: - kv  23:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:  161 tensors
llama_model_loader: - type q4_0:  225 tensors
llama_model_loader: - type q6_K:    1 tensors
llm_load_vocab: special tokens cache size = 1507
llm_load_vocab: token to piece cache size = 3,6114 MB
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = qwen2
llm_load_print_meta: vocab type       = BPE
llm_load_print_meta: n_vocab          = 92416
llm_load_print_meta: n_merges         = 112005
llm_load_print_meta: n_ctx_train      = 65536
llm_load_print_meta: n_embd           = 4096
llm_load_print_meta: n_head           = 32
llm_load_print_meta: n_head_kv        = 4
llm_load_print_meta: n_layer          = 32
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_embd_head_k    = 128
llm_load_print_meta: n_embd_head_v    = 128
llm_load_print_meta: n_gqa            = 8
llm_load_print_meta: n_embd_k_gqa     = 512
llm_load_print_meta: n_embd_v_gqa     = 512
llm_load_print_meta: f_norm_eps       = 0,0e+00
llm_load_print_meta: f_norm_rms_eps   = 1,0e-05
llm_load_print_meta: f_clamp_kqv      = 0,0e+00
llm_load_print_meta: f_max_alibi_bias = 0,0e+00
llm_load_print_meta: f_logit_scale    = 0,0e+00
llm_load_print_meta: n_ff             = 13440
llm_load_print_meta: n_expert         = 0
llm_load_print_meta: n_expert_used    = 0
llm_load_print_meta: causal attn      = 1
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = 2
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 1000000,0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_ctx_orig_yarn  = 65536
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: ssm_d_conv       = 0
llm_load_print_meta: ssm_d_inner      = 0
llm_load_print_meta: ssm_d_state      = 0
llm_load_print_meta: ssm_dt_rank      = 0
llm_load_print_meta: model type       = 7B
llm_load_print_meta: model ftype      = Q4_0
llm_load_print_meta: model params     = 7,25 B
llm_load_print_meta: model size       = 3,89 GiB (4,61 BPW) 
llm_load_print_meta: general.name     = 7b0cc3380fe815e6f08fe2f80c03e05a8b1883d8
llm_load_print_meta: BOS token        = 2 '<|endoftext|>'
llm_load_print_meta: EOS token        = 4 '<|im_end|>'
llm_load_print_meta: UNK token        = 0 '<unk>'
llm_load_print_meta: PAD token        = 92298 '<fim_pad>'
llm_load_print_meta: LF token         = 35526 'Ä'
llm_load_print_meta: EOT token        = 2 '<|endoftext|>'
llm_load_tensors: ggml ctx size =    0,18 MiB
llm_load_tensors:        CPU buffer size =  3983,84 MiB
..........................................................................................
llama_new_context_with_model: n_ctx      = 65536
llama_new_context_with_model: n_batch    = 2048
llama_new_context_with_model: n_ubatch   = 512
llama_new_context_with_model: flash_attn = 0
llama_new_context_with_model: freq_base  = 1000000,0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init:        CPU KV buffer size =  4096,00 MiB
llama_new_context_with_model: KV self size  = 4096,00 MiB, K (f16): 2048,00 MiB, V (f16): 2048,00 MiB
llama_new_context_with_model:        CPU  output buffer size =     0,35 MiB
llama_new_context_with_model:        CPU compute buffer size =  4256,01 MiB
llama_new_context_with_model: graph nodes  = 1126
llama_new_context_with_model: graph splits = 1

system_info: n_threads = 6 / 12 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | AVX512_BF16 = 0 | FMA = 1 | NEON = 0 | SVE = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | MATMUL_INT8 = 0 | LLAMAFILE = 1 | 
main: interactive mode on.
sampling: 
        repeat_last_n = 64, repeat_penalty = 1,000, frequency_penalty = 0,000, presence_penalty = 0,000
        top_k = 40, tfs_z = 1,000, top_p = 0,950, min_p = 0,050, typical_p = 1,000, temp = 0,800
        mirostat = 0, mirostat_lr = 0,100, mirostat_ent = 5,000
sampling order: 
CFG -> Penalties -> top_k -> tfs_z -> typical_p -> top_p -> min_p -> temperature 
generate: n_ctx = 65536, n_batch = 2048, n_predict = 512, n_keep = 0


== Running in interactive mode. ==
 - Press Ctrl+C to interject at any time.
 - Press Return to return control to the AI.
 - To return control without starting a new line, end your input with '/'.
 - If you want to submit another line, end your input with '\'.

You are a helpful assistant.
<?php //Create a Hook for EspoCRM for a module called Tester that checks if Entity name has length divisible by 3.
<0x0A>You[UNK_BYTE_0xe29681▁are]are[UNK_BYTE_0xe29681▁a]a[UNK_BYTE_0xe29681▁helpful]helpful[UNK_BYTE_0xe29681▁assistant]assistant.
```

Steps to reproduce the BFE pretokenizer bug:
1. Download Qwen/CodeQwen1.5-7B-Chat from huggingface
2. Run convert-hf-to-gguf.py on the model

Steps to reproduce the weird output bug:
1. Download Qwen/CodeQwen1.5-7B-Chat from huggingface
2. Replace the qwen2 chkhsh with cd88bc280b3debbdddec3304015afd7e215c61d674846a2dac7271275384810c
3. Run convert-hf-to-gguf.py on the model
4. Run ./llama-quantize <converted_model> <name_of_quantized_model> q4_0
5. Run ./llama-cli -m <name_of_quantized_model> -n 512 --color -i -f prompts/chat-with-qwen.txt

I'm not very skilled with AI, so I might be missing something important.

### Name and Version

./llama-cli --version
version: 3148 (e65bbf60)
built with cc (GCC) 14.1.1 20240522 for x86_64-pc-linux-gnu

### What operating system are you seeing the problem on?

Linux

**Comment by jarca0123** [comment]
Oops, I forgot the tokenizer.model file, which huggingface doesn't download. It works perfectly then.

**Comment by Josh-Dum** [comment]
I have the same problem. Have you found a solution ?

**Comment by Josh-Dum** [comment]
How did you resolve this problem exactly i have the same issue .

**Comment by jarca0123** [comment]
> How did you resolve this problem exactly i have the same issue .
I downloaded everything from the repo here: https://huggingface.co/Qwen/CodeQwen1.5-7B-Chat/tree/main. When I tried to run the command again using everything from the repo, it then worked as I mentioned. Anyway, when you fetch the tokenizer using the python huggingface interface, it doesn't download the tokenizer.model file, which then resulted in the issue that I mentioned.
